### ISSUE TITLE
Add a manual if possible

### ISSUE THREAD

**Issue by GabrieleNocchi** [petrelharp/local_pca#36]
Such a pity that this tool does not have a proper manual or example showing a full analysis. It is a great tool, but some stuff really take too long to understand that it was quicker to recode it myself.

I think it should be explained better how to select outliers, how to use the corners function, how to join outliers windows and do a PCA based on joined outliers windows.

I think also it should have an option to make windows based on other values, such as bp or cM. Not just SNPs number

**Comment by petrelharp** [comment]
Well, thanks for the feedbac, @GabrieleNocchi, but could you be a bit more precise with your recommendations here? Let's see:
1. As linked to in [the README](https://github.com/petrelharp/local_pca), there are three full analyses (human, medicago, and drosophila), although for your own code it's recommended to start with the code in the [templated/ directory](https://github.com/petrelharp/local_pca/blob/master/templated/README.md) - why did these not count as "an example showing a full analysis"?
2. Did you notice the help, e.g., `help(corners)`?  It's true, there's no walk-through of using these things, but there is (a) documentation, and (b) working examples?
```
Find Points in a 2D Cloud Closest to the "Corners"

Description:

     Finds the k "extreme points" that lie closest to the minimal
     enclosing circle, and returns a matrix with k columns containing
     the portion of points that lie closest to those extreme points. If
     k is larger than the number of points on the minimal enclosing
     circle, adds the closest points to that circle that are not in
     previously added points' neighborhoods until k is reached.

Usage:

     corners(xy, prop, k = 3)
     
Arguments:

      xy: A two-column numeric matrix of coordinates.

    prop: The proportion of points to return for each corner.

Value:

     A three column integer matrix giving the indices of the
     corresponding points.
```
3. There is an option to choose windows based on bp: see for instance the `type="bp"` argument; from `help(vcf_windower)`:
```
    type: The units of the window: 'bp' or 'snp'?
```

It would be nice to have windows based on cM, I agree. It wouldn't be that hard, following the "bp" example. Pull requests welcome!

**Comment by GabrieleNocchi** [comment]
Hi Peter, thanks a lot for your reply.

1. I think they do count, but those examples are a bit hard to follow in my opinion, with several scripts and also functions not in the package. But it could well be me not being a very good programmer, I admit. 
I was following the drosophilia example to try and get and join the outliers like it is described in the paper, but I could not understand some of the steps even after reading the code, or it was taking me too long that it was quicker to re-code. I think adding some more steps to the quick example here: popres/popres_example.R  would be much easier to follow .
The run_lostruct.R is a bit over complicated and hard to follow in my opinion, it does a bit too many things and I like to have more control over what is happening by seeing each step, rather than inputting my data as bcf files (which I do not like using) and running that pre made pipeline.

2. Yes, i read the documentation and raw code. But was struggling to understand for example what I need to give to corners function as input (a matrix yes -- but which?), as it was throwing errors with everything I tried. I thought it wanted the fit2d$points from the quick example.

3. Thanks! I will try tomorrow. I did not understand what that vcf_windower was doing, I thought it was just another way to read in the vcf in bcf format without keeping it all in memory (by reading home page/README of repository and quick example). I thought the function to make the windows is  eigen_windows. As the home page/README shows the 3 ways to read the data and then it goes: "In any case next step is" and uses eigen_windows.

Sorry, I did not want to sound overly critic. As I said, this is a very useful tool and I am very glad you made it. I was being critic exactly because of that, as it is a very useful and great tool and I think would benefit if it was a bit more accessible from an IT point of view!
I found my way to do what I needed (I think), but I just had an hard time following those more detailed analyses! I think building on the quick example would be great, keeping the code simple and in a single script (following the same simple R example format)!
Then it mite make easier also to follow the more complicated examples

**Comment by GabrieleNocchi** [comment]
2. In addition to the point 2. In my tries, I managed to make the corners function work (I do not remember exactly how, can give some more try tomorrow) but I struggled to also understand the output:
**A three column integer matrix giving the indices of the corresponding points.**
   
   
I could see in those 3 output indices, that one of these 3 seemed to be the corresponding row of the outlier windows in the input. But I could not understand what the other 2 indices given as value of the corners functions were (why 3 indices?). So basically I found my outliers manually with, for example:
     
  ` outliers <- which(fit2d$points[,2] < -0.4)`
     
This return some indexes (one index per outlier window, corresponding to row where the outlier window is in the input) for those points that looked as outliers in my distance plot (with Coord2 below -0.4, in my case).
Then I looked in eigenstuff, and plotted at those windows with:
     

`for (i in outliers) {`
`toplot <- eigenstuff[i,]`
`p1 <- ggplot() +`
`    geom_point(aes(x = toplot[4:383], y = toplot[384:763]))    `
`  }`
    
Just a simplified version of what I did. (my dataset has 380 individuals and I calculated PC1 and PC2, hence the numbers used in the plotting code).
    
Then I also coded another patchy part to actually join all the outliers together and do a PCA based on the SNPs of all the outliers together.

**Comment by petrelharp** [comment]
Thanks. FWIW I do agree with you - more documentation would be useful, and documentation is certainly worthwhile. Unfortunately, I've got too much on my plate to write a whole new set of docs.

Also - if you're going to start with code, I strongly recommend starting with the `templated/` code rather than the examples.

**Comment by petrelharp** [comment]
I have had a go at making the docs for `corners()` more clear - see what you think, and feel free to suggest improvements? Click the "files changed" tab at #37.

**Comment by GabrieleNocchi** [comment]
I finally understood that function. The description is a bit better, but if I have to be honest it took me some time and I had to play with the data again to understand it. Again, trying to help but it could well be me. I think it is more complicated than it needs to be -- you just want something to quickly pull some outliers.

In my opinion, the output of corners is given in a strange format which is not intuitive for me. I would probably simplify it by simply returning a vector of indices, without the columns and rows subdivision.  Just return the indices of the outliers without saving from which corners they belong to and putting them all in the same row (just save them in a vector). Also, I think you need to put some error or warnings that if the user puts a K less than 3 (in the corner function), it will still pick 3 out of corners. 

In case you prefer the function as is, I would give a simple word description of the value returned:
 
The columns are the corners, but I would change the description with (in bold changes, in italic what I would remove as it confuses me):
Finds the k "extreme points" (the "corners") that lie closest to the minimal circle enclosing the input collection of points in two dimensions, and returns for each of these 'corners' which of the input points are closest to that corner. The return value is a matrix with k columns, whose **k-th** column contains the indices of those points that lie closest to the **k-th** corner. **The minimum number of k is 3.** _Each column has (indices for) a given proportion of the input points, and so points may appear in more than one column._

 Another issue in the value in my opinion is the rows, which are a bit hard to describe - that is also what was confusing me. Because each row is basically a group of outliers (as many as the K you use), so for 3 corners each row gives 3 outliers windows (unless a point is outliers for more than one corner) and each column has the index for an outlier for each corner. That s why I think the format is confusing, each row has a K number of outliers and it is strange they are given in the same row.

If you want to keep the information of which corner each outlier was taken from, I think other data formats would be better such as a list where each list entry has outliers for a corner: my_list[1] with outliers of corner1, my_list[2] outliers corner 2 etc. But I think a simple vector as I suggested is easier, without saving which corner an outlier is taken from. Given a simple vector with the indices of outliers, the user can simply plot and highlight those points in  fit2d$points plot to see which corners each belongs to.

I would add a couple line to the quick example after the plot of fit2d$points as it is actually the reason of the whole tool (to find outliers), I think adding this would "complete" the quick example a little more if you do not have time to do a full manual/tutorial (I still think there also should be a couple extra lines to plot the PCA based on those outliers singularly and also to join them, do a PCA based on all and plot it -- but I understand time constraint users can find their way to do that as I did). I d add at a minimum something like:

`#find the 0.05 outliers from 3 corners of the plotting space:`
`outliers <- corners(fit2d$points, prop = 0.05, k = 3)`

**Comment by petrelharp** [comment]
Thanks for the comments! Let's see:
- I don't want to change `i` to `k`, because `k` is the *total* number of columns; so `i` is standing in for an arbitrary column.
- I've replaced the italicized bit with something more descriptive.
- I've added an example.

Could you have a look?

> I would add a couple line to the quick example after the plot of fit2d$points

Sorry - which file are you referring to here?

**Comment by petrelharp** [comment]
Oh, sorry - I also meant to say that your suggestion for a different format is a good one, but I don't want to change that at this point, as it would break people's code.

**Comment by GabrieleNocchi** [comment]
Yep just had a look, it is clear enough now. I understand the i to k thing, I think you are right and it is more correct with "i" like you said. I suggested the change as initially the i-th there confused me a little bit, but I understand "i" is what people usually use to loop along columns etc...

The example I am referring to is:
https://github.com/petrelharp/local_pca/blob/master/popres/popres_example.R

which is referenced in the README/Home Page of the rep with:
[A quick example](https://github.com/petrelharp/local_pca/blob/master/popres/popres_example.R) : in four lines of code, reads in chromosome 22 from a TPED, and does local PCA.

I know you said that it is better to start with the code in the templated directory. But IMO, that is way too complicated to start with and take some time to follow run_lostruct.R, at least for me. 

I think this popres_example.R linked as quick example in the README is really great, and it is the best starting point for a new user of lostruct (IMO). It has the steps in a simple format, without extra custom functions and line by line is easy to follow and uses most of the package functions -- I think without that I would have not been able to understand how to use the package in a timely manner and I mite have given up. And after that, the user can move to templated directory for the more advanced examples. 


I would add a couple of very simple lines to  popres_example.R as I said, to finish the example script using the corners function to detect and extract outliers -- and if you have time, add how to plot the outliers PCA if you have a simple one liner using output of corners (but for that, i think people can find their own way to do it if you can't).

**Comment by petrelharp** [comment]
Ah, I like that idea. I guess I don't want to just add those lines without actually seeing what it looks like, and I don't have those data any more; really I should do what you're suggesting using some publically available data so people can follow along. I've opened a new issue for this (#38) and will close this one.

Thanks!
